great_expectations/rule_based_profiler/rule_based_profiler.py

```diff
@@ -295,27 +343,250 @@ def run(
+    def reconcile_rules_for_profiler(
+        self, rules: Optional[Dict[str, Dict[str, Any]]] = None
+    ) -> List[Rule]:
+        if rules is None:
+            rules = {}
 
-        """
-        # Provide visibility into parameters that RuleBasedProfiler was instantiated with.
-        report_object: dict = {"config": self._citation}
+        effective_rules: Dict[str, Rule] = self._get_rules_as_dict()
+
+        rule_name: str
+        rule_config: dict
+
+        override_rule_configs: Dict[str, Dict[str, Any]] = {
+            rule_name: RuleBasedProfiler._reconcile_rule(
+                existing_rules=effective_rules,
+                rule_name=rule_name,
+                rule_config=rule_config,
+            )
+            for rule_name, rule_config in rules.items()
+        }
+        override_rules: Dict[str, Rule] = {
+            rule_name: self._init_one_rule(rule_name=rule_name, rule_config=rule_config)
+            for rule_name, rule_config in override_rule_configs.items()
+        }
+        effective_rules.update(override_rules)
 
-        if pretty_print:
-            print(f"\nRuleBasedProfiler class name: {self.name}")
+        return list(effective_rules.values())
 
-            if not self._variables:
-                print(
-                    'Your current RuleBasedProfiler configuration has an empty "variables" attribute. \
-                    Please ensure you populate it if you\'d like to reference values in your "rules" attribute.'
+    @staticmethod
+    def _reconcile_rule(
+        existing_rules: Dict[str, Rule], rule_name: str, rule_config: dict
+    ) -> Dict[str, Any]:
+        effective_rule_config: Dict[str, Any]
+        if rule_name in existing_rules:
+            rule: Rule = existing_rules[rule_name]
+            domain_builder_config: dict = rule_config.get("domain_builder", {})
+            effective_domain_builder_config: dict = (
+                RuleBasedProfiler._reconcile_domain_builder_config(
+                    rule=rule,
+                    domain_builder_config=domain_builder_config,
                 )
+            )
+            effective_parameter_builder_configs: Optional[
+                List[dict]
+            ] = RuleBasedProfiler._reconcile_parameter_builder_configs_for_rule(
+                rule=rule,
+                rule_config=rule_config,
+            )
+
+            effective_expectation_configuration_builder_configs: List[
+                dict
+            ] = RuleBasedProfiler._reconcile_expectation_configuration_builder_configs_for_rule(
+                rule=rule,
+                rule_config=rule_config,
+            )
+            effective_rule_config = {
+                "domain_builder": effective_domain_builder_config,
+                "parameter_builders": effective_parameter_builder_configs,
+                "expectation_configuration_builders": effective_expectation_configuration_builder_configs,
+            }
```

Comment:
Instead of `Dict[str, Any]`, should we leverage a custom type? Connects to the earlier point made.

Reply:
This is essentially `RuleConfig` right?

Reply:
I would prefer to figure this out as a separate effort.  Right now, making this a single-use type does not appear to be adding any value.

Reply:
I think the value comes in knowing the content of the type. A dictionary somewhat detracts from the knowledge gained by our config object pattern.